File: Src/java/engine/src/main/java/org/opencds/cqf/cql/engine/elm/execution/TimeFromEvaluator.java
package org.opencds.cqf.cql.engine.elm.execution;

import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument;
import org.opencds.cqf.cql.engine.execution.Context;
import org.opencds.cqf.cql.engine.runtime.DateTime;
import org.opencds.cqf.cql.engine.runtime.Precision;
import org.opencds.cqf.cql.engine.runtime.Time;

/*
time from(argument DateTime) Time

NOTE: this is within the purview of DateTimeComponentFrom
  Description available in that class
*/

public class TimeFromEvaluator extends org.cqframework.cql.elm.execution.TimeFrom {

    public static Object timeFrom(Object operand) {
        if (operand == null) {
            return null;
        }

        if (operand instanceof DateTime) {
            int hour;
            if (((DateTime)operand).getPrecision().toDateTimeIndex() > 2) {
                hour = ((DateTime)operand).getDateTime().getHour();
            }
            else {
                return null;
            }

            int minute;
            if (((DateTime)operand).getPrecision().toDateTimeIndex() > 3) {
                minute = ((DateTime)operand).getDateTime().getMinute();
            }
            else {
                return new Time(hour);
            }

            int second;
            if (((DateTime)operand).getPrecision().toDateTimeIndex() > 4) {
                second = ((DateTime)operand).getDateTime().getSecond();
            }
            else {
                return new Time(hour, minute);
            }

            int millisecond;
            if (((DateTime)operand).getPrecision().toDateTimeIndex() > 5) {
                millisecond = ((DateTime)operand).getDateTime().get(Precision.MILLISECOND.toChronoField());
            }
            else {
                return new Time(hour, minute, second);
            }

            return new Time(hour, minute, second, millisecond);
        }

        throw new InvalidOperatorArgument(
                "TimeFrom(DateTime)",
                String.format("TimeFrom(%s)", operand.getClass().getName())
        );
    }

    @Override
    protected Object internalEvaluate(Context context) {
        Object operand = getOperand().evaluate(context);
        return timeFrom(operand);
    }
}

File: Src/java/engine/src/commonMain/kotlin/org/opencds/cqf/cql/engine/elm/executing/TimeFromEvaluator.kt
package org.opencds.cqf.cql.engine.elm.executing

import kotlin.jvm.JvmStatic
import org.opencds.cqf.cql.engine.exception.InvalidOperatorArgument
import org.opencds.cqf.cql.engine.runtime.DateTime
import org.opencds.cqf.cql.engine.runtime.Precision
import org.opencds.cqf.cql.engine.runtime.Time
import org.opencds.cqf.cql.engine.util.javaClassName

/*
time from(argument DateTime) Time

NOTE: this is within the purview of DateTimeComponentFrom
  Description available in that class
*/
@Suppress("MagicNumber")
object TimeFromEvaluator {
    @JvmStatic
    fun timeFrom(operand: Any?): Any? {
        if (operand == null) {
            return null
        }

        if (operand is DateTime) {
            val hour: Int
            if (operand.precision!!.toDateTimeIndex() > 2) {
                hour = operand.dateTime!!.getHour()
            } else {
                return null
            }

            val minute: Int
            if (operand.precision!!.toDateTimeIndex() > 3) {
                minute = operand.dateTime!!.getMinute()
            } else {
                return Time(hour)
            }

            val second: Int
            if (operand.precision!!.toDateTimeIndex() > 4) {
                second = operand.dateTime!!.getSecond()
            } else {
                return Time(hour, minute)
            }

            val millisecond: Int
            if (operand.precision!!.toDateTimeIndex() > 5) {
                millisecond = operand.dateTime!!.get(Precision.MILLISECOND.toChronoField())
            } else {
                return Time(hour, minute, second)
            }

            return Time(hour, minute, second, millisecond)
        }

        throw InvalidOperatorArgument("TimeFrom(DateTime)", "TimeFrom(${operand.javaClassName})")
    }
}
